src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs

```diff
@@ -673,19 +683,48 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
         {
             BoundSpillSequenceBuilder builder = null;
             var right = VisitExpression(ref builder, node.Right);
-            BoundExpression left;
-            if (builder == null || node.Left.Kind == BoundKind.Local)
+
+            BoundExpression left = node.Left;
+            if (builder == null)
             {
-                left = VisitExpression(ref builder, node.Left);
+                left = VisitExpression(ref builder, left);
             }
             else
             {
                 // if the right-hand-side has await, spill the left
                 var leftBuilder = new BoundSpillSequenceBuilder();
-                left = VisitExpression(ref leftBuilder, node.Left);
-                if (left.Kind != BoundKind.Local)
+
+                switch (left.Kind)
                 {
-                    left = Spill(leftBuilder, left, RefKind.Ref);
+                    case BoundKind.Local:
+                    case BoundKind.Parameter:
+                        // locals and parameters are directly assignable, LHS is not on the stack so nothing to spill
+                        break;
+
+                    case BoundKind.FieldAccess:
+                        var field = (BoundFieldAccess)left;
+                        // static fields are directly assignable, LHS is not on th estack, nothing to spill
+                        if (field.FieldSymbol.IsStatic) break;
+
+                        // instance fields are directly assignable, but receiver is pushed, so need to spill that.
+                        var receiver = VisitExpression(ref leftBuilder, field.ReceiverOpt);
+                        receiver = Spill(builder, receiver, field.FieldSymbol.ContainingType.IsValueType ? RefKind.Ref : RefKind.None);
+                        left = field.Update(receiver, field.FieldSymbol, field.ConstantValueOpt, field.ResultKind, field.Type);
+                        break;
+
+                    case BoundKind.ArrayAccess:
+                        var arrayAccess = (BoundArrayAccess)left;
+                        // array and indices are pushed on stack so need to spill that
+                        var expression = VisitExpression(ref leftBuilder, arrayAccess.Expression);
+                        expression = Spill(builder, expression, RefKind.None);
+                        var indices = this.VisitExpressionList(ref builder, arrayAccess.Indices, forceSpill: true);
+                        left = arrayAccess.Update(expression, indices, arrayAccess.Type);
+                        break;
+
+                    default:
+                        // must be indirectly assignable, just visit and spill
```

Comment:
In that case the LHS has sideeffects, so it would need to be evaluated as a whole and then the result spilled. Which in a case of a ref returning method will fail. 

I.E. 
```
obj,Method() = await Something();
```

In this case we cannot decompose the LHS into parts from which the LHS could be rehydrated back after await. 
Unlike assignments to locals, fields and arrays, method calls are opaque variables. When they are byref we cant spill them.


---
In reply to: [127502849](https://github.com/dotnet/roslyn/pull/20883#discussion_r127502849) [](ancestors = 127502849)